source/include/localisation/cs_TrackableEvent.h

```diff
@@ -11,6 +11,25 @@
 #include <localisation/cs_TrackableId.h>
 
 // REVIEW: Why a derivative of event_t, instead of a data struct?
+// @bart: I think this makes the user code much cleaner.
+/**
+ * E.g.
+ *
+ * Now it's just:
+ * 	TrackableEvent trackEvt;
+ * 	trackEvt.rssi = scannedDevice->rssi;
+ * 	trackEvt.dispatch();
+ *
+ * Rather than:
+ * trackable_event_data_t trackdata;
+ * trackdata.rssi = scannedDevice->rssi;
+ * event_t trackevent(CS_TYPE::EVT_TRACKABLE, &trackdata, sizeof(trackdata);
+ * trackevent.dispatch();
+ *
+ * Not only saving a line of non-trivial boilerplate at each call site, but also
+ * reducing the surface at which we have to do size checks to the event class
+ * implementations rather than al call sites.
+ */
```

Comment:
It's different from all other events, so that's confusing.

What's confusing me the most is how in `NearestCrownstoneTracker::handleEvent` you cast `event.data` to `TrackableEvent`, so a derivative of an `event_t` ended up in the data of an `event_t`?

I don't see any change in the number size checks.

Reply:
The inheritance and construction are to facilitate ease of use (as explained by the example) the fact that it's usage on the dispatching side is different from other events is (IMO) for the better. If all events were like this the event system would already be a lot safer to use. :-)

Explaining the mechanics:
Effectively TrackableEvent is playing the role of trackable_event_data_t. TrackableEvent is constructed with `this` as the `data` parameter of its parents constructor and `CS_TYPE::EVT_TRACKABLE` as its type. Thus when the `::handleEvent` function receives an `event_t` of that type, it knows the data should be cast to `TrackableEvent`. This is no else than any other cast in a handler we're making and is documented exactly the same as all others in cs_Types.h.

Reply:
There are 73 `EVT_XXX` types defined in `cs_Types.h`. @ArrowAcrobatics Do you recommend to create `cs_XxxEvent.h` and `cs_XxxEvent.cpp` for each of them? That's 146 files.

Reply:
Yes. 

This refactor eventually will safe program size when events that are not used can be eliminated by putting event description stuff in the folder of their modules behind the relevant `BUILD_YYY` flags. We can inline the definition of the constructor, avoiding the *2 multiplier, and perhaps sort them by theme to reduce it to 1 file per module, but overall I'm very much pro splitting up the humongous cs_Types into something that is maintainable. 

Increasing locality of information makes it possible to just copy paste a template file rather than having to edit 12 different locations before you have a new event implemented too.

Reply:
I like that it's 1 point where you avoid a type error, though right now it's simply moved to this class. Maybe such classes can be templated?

I still don't see why it has to be derived from `event_t`, dragging in a lot of stuff. You could also give it a dispatch() function with something like:
```
event_t event;
event(CS_TYPE::EVT_TRACKABLE, this, sizeof(this))
event.dispatch();
```
right?

Another thing i don't like is that it's a full class, instead of a struct. I'd rather have the data struct separated from these convenience classes.

Reply:
I think `this` doesn't mean what you expect in your example. It refers to an object of the type equal to the calling scope.

@vliedel All of the event_t stuff in the calling scope is used in your example too, inheritance doesn't add anything extra there. It's just a code-reuse pattern. I'm just using it to reduce the dispatching user-side code.

Templating and extracting data structs from the events is of course possible to further reduce boiler plate stuff, but the code as in the commit is already a step towards that. It's tempting to start overcomplicating before implementing.
I've added an example piece of code to this comment to show what the templated way _might_ look like.

[GenericEvents.txt](https://github.com/crownstone/bluenet/files/6030060/GenericEvents.txt)

Reply:
@ArrowAcrobatics I understand what you're after. If there's a type kept by scope it might be maintainable. 

The reason the types are centralized right now is as follows:

- There is a separation between the POD which is preferably **compressed** (e.g. packed) and preferably **used everywhere** for (1) compact storage, (2) compact event communication within bluenet, (3) compact IPC between bluenet and microapp, and (4) compact communication over bluetooth. Each time there's conversion required somewhere, we will need flash and ram. We've to find some common ground of truth for data.
- If there are PODs it is very logical to get convenience functions or classes operating on all of them. This is what happened with all the functions that check things at compile time.

I don't mind if it is restructured completely. As long as there's a single data representation and not yet another one for the event bus. I wouldn't like serialize/deserialize types of operators everywhere. Sure, fds storage, characteristic buffers, microapp buffers, etc. are all different, but in the end they all profit from small compact representations.

The `BUILD_XXX` argument is a bit void. We never keep them at `=0` forever. So, it will end up in the firmware. 

You will have lists of `#include <events/cs_TrackableEvent.h>`  in files that handle many events. I foresee quite a few `#if BUILD_XXX` statements in all the files that will need the events. 

On the other hand, it is nice that on changing the data structure of an event, only the relevant files have to be recompiled.

Reply:
Agreed: PODs need to be defined once for a data type and reused whereever possible. (E.g. use the same POD on event bus as over UART)

I think that by templating it we can ensure that PODs are only included in the source code when there is at least one user (either receiving or dispatching) as well as ensuring that when neither exists, the POD definition is not included in the generated binary. In this setting we only need to have a single `BUILD_XXX` flag per component.

Reply:
I suggest that until we agree on the better way, we use the old way, especially for this pull request.

Regarding the `BUILD_` settings, those are usually temporary, for modules that are not release ready yet.
Regarding the `event_t` data, i think you now send an `event_t` with payload of type `event_t`, so you have all `event_t` fields double.

Reply:
removed in https://github.com/crownstone/bluenet/commit/9e6e1f2393dca7631c614b0ec43c67cb63bcf69f